finetune/nn/target_blocks.py

```diff
@@ -517,6 +517,185 @@ def seq_lab_internal(hidden):
+                    input_tensor=one_hot_class_weights, axis=-1, keepdims=True
+                )
+                logits = class_reweighting(per_token_weights)(logits)
+                                                                                                      
+            transition_params = tf.cast(
+                tf.compat.v1.get_variable(
+                    "Transition_matrix", shape=[n_targets, n_targets]
+                ),
+                tf.float32
+            )
+            if targets is not None:
+                # Better way to do this?
+                featurizer_fn = kwargs.get("featurizer_fn")
+
+                def after_embeddings(embeddings):
+                    featurizer_state = featurizer_fn(embeddings)
+                    hidden = featurizer_state["sequence_features"]
+                    logits = layer(hidden)
+                    return tf.cast(logits, tf.float32)
+                def after_transformer(hidden):
+                    logits = layer(hidden)
+                    return tf.cast(logits, tf.float32)
+                out_fn = after_embeddings if featurizer_fn else after_transformer
+                preturb_target = embeddings if featurizer_fn else hidden
+
+                target_shape = tf.shape(targets)
+                all_logits = logits
+                all_lengths = lengths
+                logits = all_logits[:target_shape[0], :target_shape[1]]
+                lengths = all_lengths[:target_shape[0]]
+
+                if use_crf:
+                    log_likelihood, _ = crf_log_likelihood(
+                        logits, targets, lengths, transition_params=transition_params
+                    )
+                    loss = -log_likelihood
+                else:
+                    weights = tf.sequence_mask(
+                        lengths, maxlen=tf.shape(input=targets)[1], dtype=tf.float32
+                    ) / tf.expand_dims(tf.cast(lengths, tf.float32), -1)
+                    loss = tf.compat.v1.losses.sparse_softmax_cross_entropy(
+                        targets,
+                        logits,
+                        weights=weights
+                    )
+
+                    e = 0.02
+                    k = 1
+                    probs = tf.stop_gradient(tf.nn.softmax(all_logits))
+                    adv_vector = tf.random.uniform(shape=tf.shape(preturb_target),
+                                                   dtype=tf.float32)
+                    adv_vector = e * tf.nn.l2_normalize(adv_vector, axis=-1)
+                    batch_size = tf.cast(tf.shape(preturb_target)[0], tf.float32)
+                    kl_div = tf.losses.KLDivergence(reduction=tf.keras.losses.Reduction.NONE)
+                    mask = tf.sequence_mask(all_lengths,
+                                            maxlen=tf.math.reduce_max(all_lengths),
+                                            dtype=tf.float32)
+                    for _ in range(k):
```

Comment:
I believe starting with k = 0 should give you something functional albeit without an optimal adv_vector